### DISCUSSION TITLE
How to dump to a typed dict?

### DISCUSSION THREAD

**Discussion by PelicanQ** [pydantic/pydantic#8093] [Question]
I find 0 other people asking this so I just have to ask:

**Q**: How could I get a typed "dump" from a Pydantic model?

**My context**:

We have `.model_dump()` but its return type hint is `dict[str, Any]`. This is annoying is because I want to write a FastAPI backend with SQLAlchemy ORM and Pydantic models. When creating an ORM model we have only one option (I think) to create an instance, calling constructor with kwargs. Say `User` is an SQLAlchemy model (declarative base), we create one like:

`User(name="John")`

With a Pydantic model `UserCreate` we could feed it in like 

```py
user = UserCreate(name="John")
User(**user.model_dump())
```

Now we cannot get static type checking to make sure our Pydantic UserCreate fields align with SQLAlchemy User model.
So if I add a new field to my SQLAlchemy model like:

```py
from sqlalchemy.orm import MappedAsDataclass, DeclarativeBase

class BaseDBModel(MappedAsDataclass, DeclarativeBase):
    pass

class User(BaseDBModel):
    __tablename__ = "users"
    id: Mapped[int] = mapped_column(primary_key=True, init=False)
    name: Mapped[str] = mapped_column()
    age: Mapped[int] = mapped_column()
```
 and my PydanticModel is

```py
from pydantic import BaseModel

class UserCreate(BaseModel):
    name: str
```
There's no way for a static type checker to see that the dict `user.model_dump()` is missing an `age`.
```py
user = UserCreate(name="John")
User(**user.model_dump())
```

so it's left to runtime errors to realize you forgot `age` on `UserCreate`. I know Tiangolo (creator of FastAPI) has his own ORM (SQLModel) which might fix this, but it's too young to adopt.

SQLAlchemy version 2.0.23
Pydantic version 2.5

**Comment by Hugo-Polloli** [comment]
The issue is even more general than just Pydantic + an ORM
We're using only Pydantic, like you we have a create user DTO (UserCreate) and an "internal" complete model (User), whether the class you want to build from UserCreate  is a basic python, pydantic or orm class, you won't have typing, and you need to implement mappers or constructors :/

**Comment by gyanu2507** [comment]
`model_dump()` is typed as `dict[str, Any]` on purpose. A dict cannot carry the field names, so mypy will not notice a missing `age` if you splat it into `User(**...)`.

Pass the fields by name instead:

```python
user = UserCreate(name="John")
User(name=user.name, age=user.age)
```

If `UserCreate` has no `age`, that line is a type error. That is the check you wanted.

A helper keeps it in one place:

```python
def user_from_create(data: UserCreate) -> User:
    return User(name=data.name, age=data.age)
```

`TypeAdapter` / a `TypedDict` dump does not help here. The dump is still an untyped mapping. SQLModel is the other route if you want one class for both layers.

**Comment by nanisadw3** [comment]
Hi! 👋

While `model.model_dump()` returns a generic `dict[str, Any]` at runtime, you can ensure static type safety for your `TypedDict` in one of two clean ways:

### 1. Using `typing.cast` (Zero runtime overhead)
If you just want your IDE, mypy, or pyright to recognize the dumped dictionary as the `TypedDict`:

```python
from typing import TypedDict, cast
from pydantic import BaseModel

class UserDict(TypedDict):
    id: int
    username: str

class User(BaseModel):
    id: int
    username: str

user = User(id=1, username="inaki")
dumped_dict = cast(UserDict, user.model_dump())
```

### 2. Using `TypeAdapter` (Full runtime validation)
If you also want runtime validation ensuring the dumped dict conforms to the `TypedDict` structure:

```python
from typing import TypedDict
from pydantic import BaseModel, TypeAdapter

class UserDict(TypedDict):
    id: int
    username: str

class User(BaseModel):
    id: int
    username: str

user = User(id=1, username="inaki")
adapter = TypeAdapter(UserDict)

# Validates and types the output as UserDict
dumped_dict: UserDict = adapter.validate_python(user.model_dump())
```
